persist/fs/commitlog/reader.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		return timeZero, 0, 0, err
+	}
+
+	r.chunkReader.reset(fd)
+	r.info = schema.CommitLogInfo{}
+	if err := r.read(&r.info); err != nil {
+		r.Close()
+		return timeZero, 0, 0, err
+	}
+
+	start := time.Unix(0, r.info.Start)
+	duration := time.Duration(r.info.Duration)
+	index := int(r.info.Index)
+	return start, duration, index, nil
+}
+
+func (r *reader) Read() (
+	series m3db.CommitLogSeries,
+	datapoint m3db.Datapoint,
+	unit xtime.Unit,
+	annotation m3db.Annotation,
+	resultErr error,
+) {
+	if err := r.read(&r.log); err != nil {
+		resultErr = err
+		return
+	}
+
+	if len(r.log.Metadata) != 0 {
+		if err := proto.Unmarshal(r.log.Metadata, &r.metadata); err != nil {
+			resultErr = err
+			return
+		}
+		r.metadataLookup[r.log.Idx] = m3db.CommitLogSeries{
+			UniqueIndex: r.log.Idx,
+			ID:          r.metadata.Id,
+			Shard:       r.metadata.Shard,
+		}
+	}
+
+	metadata, ok := r.metadataLookup[r.log.Idx]
+	if !ok {
+		resultErr = errCommitLogReaderMissingLogMetadata
+		return
+	}
+
+	series = metadata
+	datapoint = m3db.Datapoint{
+		Timestamp: time.Unix(0, r.log.Timestamp),
+		Value:     r.log.Value,
+	}
+	unit = xtime.Unit(byte(r.log.Unit))
+	annotation = r.log.Annotation
+	return
+}
+
+func (r *reader) read(message proto.Message) error {
+	// Read size of message
+	size, err := binary.ReadUvarint(r.chunkReader)
+	if err != nil {
+		return err
+	}
+
+	// Extend buffer as necessary
+	if len(r.dataBuffer) < int(size) {
+		diff := int(size) - len(r.dataBuffer)
+		for i := 0; i < diff; i++ {
+			r.dataBuffer = append(r.dataBuffer, 0)
+		}
+	}
+
+	// Size the target buffer for reading and unmarshalling
+	buffer := r.dataBuffer[:size]
+
+	// Read message
+	if _, err := r.chunkReader.Read(buffer); err != nil {
+		return err
+	}
+	if err := proto.Unmarshal(buffer, message); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *reader) Close() error {
+	if r.chunkReader.fd == nil {
+		return nil
+	}
+
+	if err := r.chunkReader.fd.Close(); err != nil {
+		return err
+	}
+
+	r.chunkReader.fd = nil
+	r.metadataLookup = make(map[uint64]m3db.CommitLogSeries)
+	return nil
+}
+
+type chunkReader struct {
+	fd        *os.File
+	buffer    *bufio.Reader
+	remaining int
+	charBuff  []byte
+}
+
+func newChunkReader(bufferLen int) *chunkReader {
+	return &chunkReader{
+		buffer:   bufio.NewReaderSize(nil, bufferLen),
+		charBuff: make([]byte, 1),
+	}
+}
+
+func (r *chunkReader) reset(fd *os.File) {
+	r.fd = fd
+	r.buffer.Reset(fd)
+	r.remaining = 0
+}
+
+func (r *chunkReader) readHeader() error {
+	header, err := r.buffer.Peek(chunkHeaderLen)
+	if err != nil {
+		return err
+	}
+
+	sizeStart, sizeEnd := 0, chunkHeaderSizeLen
+	checksumSizeStart, checksumSizeEnd := sizeEnd, sizeEnd+chunkHeaderSizeLen
+	checksumDataStart, checksumDataEnd := checksumSizeEnd, checksumSizeEnd+chunkHeaderChecksumDataLen
+	size := endianness.Uint32(header[sizeStart:sizeEnd])
+	checksumSize := endianness.Uint32(header[checksumSizeStart:checksumSizeEnd])
```

Comment:
can probably switch to use `digest.ToBuffer(header[checkSumSizeStart:]).ReadDigest()`